staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go

```diff
@@ -134,10 +136,29 @@ func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	location.Scheme = "https"
 	rloc, err := r.serviceResolver.ResolveEndpoint(handlingInfo.serviceNamespace, handlingInfo.serviceName, handlingInfo.servicePort)
 	if err != nil {
-		klog.Errorf("error resolving %s/%s: %v", handlingInfo.serviceNamespace, handlingInfo.serviceName, err)
+		utilruntime.HandleError(fmt.Errorf("error resolving %s/%s: %w", handlingInfo.serviceNamespace, handlingInfo.serviceName, err))
 		proxyError(w, req, "service unavailable", http.StatusServiceUnavailable)
 		return
 	}
+
+	// If the ServiceResolver also implements ServerNameResolver, check to see if a service's external name should be used for TLS config.
+	if serverNameResolver, ok := r.serviceResolver.(ServerNameResolver); ok {
+		serverName, isExternal, err := serverNameResolver.ResolveServerName(handlingInfo.serviceNamespace, handlingInfo.serviceName)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("error resolving ServerName %s/%s: %w", handlingInfo.serviceNamespace, handlingInfo.serviceName, err))
+			proxyError(w, req, "server name unavailable", http.StatusServiceUnavailable)
+			return
+		}
+		if isExternal {
+			handlingInfo.transportConfig.TLS.ServerName = serverName
+			handlingInfo.proxyRoundTripper, handlingInfo.transportBuildingError = transport.New(handlingInfo.transportConfig)
```

Comment:
the transport creation is done in line 219 `func (r *proxyHandler) updateAPIService(apiService *apiregistrationv1api.APIService) {`

this is overriding the config for each request, should not this be done only once, when the service is added?

cc: @liggitt

Reply:
@aojea the service resolver is also called for every request, so i followed the same pattern since this is an optional interface that can be implemented by the service resolver. do you have any other suggestions on how to structure this?

Reply:
Agree with @aojea, we should not be reconstructing the transport on every request handle. The inputs should only change as fast as the APIService and DNS name of the Service object change

Reply:
@aojea @liggitt ok, i'll see if i can figure out a way to restructure things. thanks for the review!

Reply:
@aojea @liggitt i added a simple change to check the current value of the servername before updating it. let me know if that addresses your concerns. if not, i can attempt to figure out another approach.